### PR TITLE
[SPARKR][SPARK-10711] Do not assume spark.submit.deployMode is always set

### PR DESCRIPTION
**Pull request by falaki**
In `RUtils.sparkRPackagePath()` we 
1. Call `sys.props("spark.submit.deployMode")` which returns null if `spark.submit.deployMode` is not suet
2. Call `sparkConf.get("spark.submit.deployMode")` which throws `NoSuchElementException` if `spark.submit.deployMode` is not set. This patch simply passes a default value ("cluster") for `spark.submit.deployMode`.

cc @rxin

### CODE REVIEW

**Comment by falaki** [comment]
cc @shivaram

**Comment by shivaram** [comment]
Any reason this would not be set ? My assumption was that all spark-submit applications had this -- so I guess this is for applications not using spark-submit ? 

cc @sun-rui who added the function

**Comment by falaki** [comment]
Yes, if an application starts the JVM manually (not using spark-submit) this call will fail. The change just adds safety without compromising functionality or correctness.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ private[spark] object RUtils {
         (sys.props("spark.master"), sys.props("spark.submit.deployMode"))
       } else {
         val sparkConf = SparkEnv.get.conf
-        (sparkConf.get("spark.master"), sparkConf.get("spark.submit.deployMode"))
+        (sparkConf.get("spark.master"), sparkConf.get("spark.submit.deployMode", "cluster"))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: if anything the default deploy mode should be `client` to be consistent with assumptions in the code elsewhere

**Comment by sun-rui** [comment]
@falaki , are you using RStudio, where spark-submit is not involved?

I don't think we can simply set a default value. Because it is possible that the default mode does not match spark master. For example, if the spark master is 'yarn-cluster', the default mode is "client", that does not match.

The correct policy is:
  if the master does not contain deploy mode information, the deploy mode is "client",
  else get deploy mode from the master URL.

I think this issue is not related to SparkR only. @andrewor14, any policy when handling default configurations when a spark-application is not launched via spark-submit? Should we extract some logic from spark-submit so that the logic can be called in non-spark-submit cases to keep consistency?

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
> if the master does not contain deploy mode information, the deploy mode is "client",
else get deploy mode from the master URL.

We're actually deprecating the master URLs `yarn-client` and `yarn-cluster` (#8385). In general I find it pretty confusing to have the deploy mode embedded in the master URL.

> I think this issue is not related to SparkR only. @andrewor14, any policy when handling default configurations when a spark-application is not launched via spark-submit? Should we extract some logic from spark-submit so that the logic can be called in non-spark-submit cases to keep consistency?

We do have the launcher library, but beyond that I don't think this is something we support because it gets pretty difficult to maintain. AFAIK deploy mode is the only such config so this should be fine.

**Comment by andrewor14** [comment]
LGTM merging once tests pass

**Comment by rxin** [comment]
I've merged this.

**Pull request closed without merging**